inference-engine/src/preprocessing/ie_preprocess_gapi_kernels.cpp

```diff
@@ -42,572 +42,547 @@
+struct typelist {};
 
-        if (std::is_same<T, float>::value && chs == 2) {
-            avx::mergeRow_32FC2(reinterpret_cast<const float*>(ins[0]),
-                                reinterpret_cast<const float*>(ins[1]),
-                                reinterpret_cast<float*>(out), length);
-            return;
-        }
+template<typename type_list>
+struct head;
 
-        if (std::is_same<T, float>::value && chs == 3) {
-            avx::mergeRow_32FC3(reinterpret_cast<const float*>(ins[0]),
-                                reinterpret_cast<const float*>(ins[1]),
-                                reinterpret_cast<const float*>(ins[2]),
-                                reinterpret_cast<float*>(out), length);
-            return;
-        }
+template<template<typename ...> class list, typename head_t, typename ... types>
+struct head<list<head_t, types...>> { using type = head_t;};
 
-        if (std::is_same<T, float>::value && chs == 4) {
-            avx::mergeRow_32FC4(reinterpret_cast<const float*>(ins[0]),
-                                reinterpret_cast<const float*>(ins[1]),
-                                reinterpret_cast<const float*>(ins[2]),
-                                reinterpret_cast<const float*>(ins[3]),
-                                reinterpret_cast<float*>(out), length);
-            return;
-        }
-    }
-#endif  // HAVE_AVX2
+template<typename typelist>
+using head_t = typename head<typelist>::type;
 
-#ifdef HAVE_SSE
-    if (with_cpu_x86_sse42()) {
-        if (std::is_same<T, uint8_t>::value && chs == 2) {
-            mergeRow_8UC2(ins[0], ins[1], out, length);
-            return;
-        }
+template<typename type>
+struct type_to_type {};
 
-        if (std::is_same<T, uint8_t>::value && chs == 3) {
-            mergeRow_8UC3(ins[0], ins[1], ins[2], out, length);
-            return;
-        }
+template <typename typelist>
+struct type_dispatch_impl;
 
-        if (std::is_same<T, uint8_t>::value && chs == 4) {
-            mergeRow_8UC4(ins[0], ins[1], ins[2], ins[3], out, length);
-            return;
-        }
+//FIXME: add test for type_dispatch
+template <template<typename ...> class typelist, typename... type>
+struct type_dispatch_impl<typelist<type...>> {
+    template <typename result_t, typename default_t, typename type_id_t, typename type_to_id_t, typename type_to_value_t>
+    static result_t dispatch(type_id_t type_id, type_to_id_t&& type_to_id, type_to_value_t&& type_to_value, default_t default_value) {
+        result_t res = default_value;
 
-        if (std::is_same<T, float>::value && chs == 2) {
-            mergeRow_32FC2(reinterpret_cast<const float*>(ins[0]),
-                           reinterpret_cast<const float*>(ins[1]),
-                           reinterpret_cast<float*>(out), length);
-            return;
-        }
+        bool matched = false;
+        std::initializer_list<int> ({
+            !matched && (type_id == type_to_id(type_to_type<type>{})) ?
```

Comment:
looks like it calls `operator()` for type `typed_merge_row ` which expects two template params:
`emplate<typename type, typename tag = isa_tag_t>`

the first type can be deducted, but the second `tag` will always be default here `tag=isa_tag_t`

and `std::is_same<tag, scalar_tag>` won't be true always